### ISSUE TITLE
step size for monotonically increasing values

### ISSUE THREAD

**Issue by maiermic** [ericniebler/range-v3#723]
I'm looking for a view that does the same as [boost::irange][irange]:

```c++
boost::irange(3, 14, 5) // 3 8 13
```

I only found `ints` and `iota`. None of them takes a `step_size`. It is always `1`. Have I overlooked something or is this feature missing?

[irange]: http://www.boost.org/doc/libs/1_60_0/libs/range/doc/html/range/reference/ranges/irange.html

**Comment by TiagoRabello** [comment]
I'm not entirely sure if it's either the only or the best way but you can achieve the same behavior by composing `view::ints` and `view::stride`:

    view::ints(3, 14) | view::stride(5) // 3 8 13

**Comment by gnzlbg** [comment]
Or `view::linear_distribute(3, 13, 3) ` but I think `ints` + `stride` is clearer.

**Comment by ericniebler** [comment]
Yes, the `stride` adaptor is currently how to do this. It would probably be more efficient for `ints` and `iota` to take a stride parameter directly, but I haven't been so motivated to add support for that.

**Comment by maiermic** [comment]
Thanks, is there also a recommended way to iterate from top to bottom (for example `13 8 3`) and for floating point numbers (for example `0.1 0.2 0.3`?

**Comment by gnzlbg** [comment]
@maiermic

> is there also a recommended way to iterate from top to bottom

```c++
#include <iostream>
#include <range/v3/all.hpp>
int main() {
   using namespace ranges;
   auto rng = view::ints(3, 14) | view::stride(5) | view::reverse;
   std::cout << view::all(rng) << std::endl;  // prints: 13 8 3
}
```

> for floating point numbers (for example 0.1 0.2 0.3?

```c++
#include <iostream>
#include <range/v3/all.hpp>

int main() {
   using namespace ranges;
   auto rng = view::linear_distribute(0.1, 0.3, 3);
   std::cout << view::all(rng) << std::endl;  // prints: 0.1 0.2 0.3
}
```

**Comment by maiermic** [comment]
@gnzlbg Thanks, if I iterate from top `14` to bottom `3` with step size `5`, I expect `14 9 4` instead of `13 8 3`.

**Comment by gnzlbg** [comment]
@maiermic 

> Thanks, if I iterate from top 14 to bottom 3 with step size 5, I expect 14 9 4 instead of 13 8 3.

Oh sorry, you wrote 13 8 3 before. Do you have any expectation about what the syntax for getting 14 9 4 from a `view::ints(3, 14)` should look like? (maybe form another programming language?)

There are many ways to do that in range-v3 (e.g. `view::closed_ints(3, 14) | view::drop(1) | view::stride(5) | view::reverse;`) but I'd like to know if you have something else in mind.

---

@ericniebler 

`view::closed_ints` is deprecated (and properly raises the deprecation warning) but `view::ints` is not. I honestly don't remember what the plan there was.

**Comment by maiermic** [comment]
@gnzlbg 

> Oh sorry, you wrote 13 8 3 before.

Sorry for the misunderstanding. `13 8 3` is an example of monotonically decreasing values. However, you can use different boundaries to generate this sequence. For example, you can count down from `13` to `2` or to `1` with step size `5`.

> Do you have any expectation about what the syntax for getting `14 9 4` from a `view::ints(3, 14)` should look like? (maybe form another programming language?)

Python has [`range(start, stop[, step])`](https://docs.python.org/3/library/functions.html#func-range). `range(0, -10, -2)` generates `0 -2 -4 -6 -8`. In the current version of range-v3 `view::ints(start, stop)` restricts `start` to be lower than `stop`. This restriction does not exist in Python's `range`. However, the default step size is `+1`. That's why `range(14, 3)` is empty. You have to define a negative step size to get monotonically decreasing values. If `view::ints(start, stop)` allows monotonically decreasing values (`view::ints(3, 1)` resulting in `3 2 1`) then you could use

```c++
view::ints(14, 3) | view::stride(-5) // 14 9 4
```

> There are many ways to do that in range-v3 (e.g. `view::closed_ints(3, 14) | view::drop(1) | view::stride(5) | view::reverse;`) but I'd like to know if you have something else in mind.

If you replace the numbers with variables `start`, `stop` and `step` in your example to generate monotonically decreasing values you get

```
view::closed_ints(stop, start) | view::drop(1) | view::stride(step) | view::reverse;
```

This does not work in general. I expect `14 9 4` for

```c++
int start = 14, stop = 3, step = 5;
```

and 

```c++
int start = 14, stop = 2, step = 5;
```

but the later results in `13 8 3`

**Comment by CaseyCarter** [comment]
> `view::ints | view::stride` should ideally generate the same code than putting the stride inside the `view::ints` itself. Any idea why it doesn't? (I never bothered to check).

`view::stride` is complicated by the fact that iterators that differ by `n` must adapt underlying iterators that differ by `n * stride`, yet the underlying range is not required to have a length that is a multiple of `stride`. We need to somehow track how far the underlying iterator of the stride view's end iterator is from the penultimate iterator so that when a user decrements it we back up the proper number of steps.

Currently, the `view::stride` implementation solves this problem by storing an offset indicating that this iterator is `stride - offset` steps from the previous iterator; the offset is always `0` for non-end iterators. The catch is that the offset is stored in an `atomic` so that stride views can be `const`-iterable, violating "don't pay for what you don't use."

**Comment by gnzlbg** [comment]
> The catch is that the offset is stored in an atomic so that stride views can be const-iterable, violating "don't pay for what you don't use."

I guess this will remain this way till #385 is resolved.

**Comment by gnzlbg** [comment]
FWIW range-v3 and the STL2 are in godbolt so this is `ints | stride` vs `linear_distribute`: https://godbolt.org/g/WRvJwa

**Comment by cor3ntin** [comment]
I was thinking of proposing a step parameter for `std::ranges::iota_view` (following discussions about https://wg21.link/p1894r0 ) - seems like a good addition to ranges-v3, i will submit a PR at some point
